aleph-client/src/contract/mod.rs

```diff
@@ -0,0 +1,144 @@
+/// Utilities for writing contract wrappers.
+///
+/// The functions in this module simplify parsing the data types returned by contract calls.
+pub mod util;
+
+use std::{
+    fmt::{Debug, Formatter},
+    fs::File,
+};
+
+use ac_primitives::ExtrinsicParams;
+use anyhow::{anyhow, Context, Result};
+use contract_metadata::ContractMetadata;
+use contract_transcode::{ContractMessageTranscoder, Value};
+use ink_metadata::{InkProject, MetadataVersioned};
+use serde_json::{from_reader, from_str, from_value, json};
+use sp_core::{crypto::AccountId32, Pair};
+use substrate_api_client::{compose_extrinsic, GenericAddress, XtStatus};
+
+use crate::{try_send_xt, AnyConnection, Connection, SignedConnection};
+
+/// Represents a contract instantiated on the chain.
+pub struct ContractInstance {
+    address: AccountId32,
+    ink_project: InkProject,
+}
+
+impl ContractInstance {
+    const MAX_READ_GAS: u64 = 500000000000u64;
+    const MAX_GAS: u64 = 10000000000u64;
+
+    /// Creates a new contract instance under `address` with metadata read from `metadata_path`.
+    pub fn new(address: AccountId32, metadata_path: &str) -> Result<Self> {
+        Ok(Self {
+            address,
+            ink_project: load_metadata(metadata_path)?,
+        })
+    }
+
+    /// The address of this contract instance.
+    pub fn address(&self) -> &AccountId32 {
+        &self.address
+    }
+
+    /// The metadata of this contract instance.
+    pub fn ink_project(&self) -> &InkProject {
+        &self.ink_project
+    }
+
+    /// Reads the value of a read-only, 0-argument call via RPC.
+    pub fn contract_read0(&self, conn: &Connection, message: &str) -> Result<Value> {
```

Comment:
I would say, that all these methods should be generic over `AnyConnection` instead of receiving concrete type.